### PR TITLE
Mapml vector tiles - early test

### PR DESCRIPTION
**Pull request by aaime**
This is a PR allowing for early test of vector tiling with WMS (no caching yet, and fixed styling).
In order to test a layer set it up to be both tiled and using features:

image

With the current code, the boundaries are normally drawn properly but the fill is broken.
I've initially tried to follow the rules we discussed by mail, but they systematically led to bits missing on the sides and the fill was broken anyways and so... I've decided to stop bleeding hours trying to figure out the "rules of the game" and ask for some verification instead. I'd be also happy to receive written-down, firm rules on how a polygon boundary with hidden sides should be represented, and then adhere to that in code and tests (as indicated, the ones discussed by mail do not seem to be working properly).

Here are some examples of misbehavior from GeoServer demo layers:

``tasmania_state_boundaries``

image
 
``giant_polygon`` (zoom in once)

image

``restricted`` (zoom in once)

image

``countries`` (zoom in once):

image

This last one is the only case where I see sides that don't belong pop up. I've tried to restrict the layer using a CQL_FILTER, ``&CQL_FILTER=ADMIN='Croatia'``, and checked the tile contents, I don't see ordinates going up to the north pole...

The code has comments as to what I've tried so far and the bits that I've removed are just commented out, for ease of testing.

@prushforth could you have a look?
Btw, if you do, I'd recommend disabling HTTP caching and work with the developer tools open, with cache disabled there too, for good measure.

# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [ ] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [ ] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [ ] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [ ] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by prushforth** [comment]
First quick comment:  if you remove the <map-span> elements and take the text value of the coordinate string, it should parse as coordinates.  That means that 

```html
<map-span class="bbox">-22.5 78.75 -11.25 78.75 -11.25 90</map-span> -11.25 90 -22.5 90<map-span class="bbox">-22.5 90 -22.5 78.75</map-span>
```

will parse to: -22.5 78.75 -11.25 78.75 -11.25 90 -11.25 90 -22.5 **90-22.5 90** -22.5 78.75

and the result is math errors.

**Comment by aaime** [comment]
@prushforth I find that surprising, is there some general HTML rule that this behavior is based on?
In GML spaces are irrelevant for example (but that's proper XML).

**Comment by prushforth** [comment]
I believe so textContent is what we are relying on. https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent 

Will get back to it in a minute but I think there's a bug in the viewer rendering besides that issue.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLGenerator.java [hunk 1]
```diff
@@ -291,22 +318,70 @@ private org.geoserver.mapml.xml.Point buildPoint(org.locationtech.jts.geom.Point
...
     }
+
+    private List<Object> buildTaggedLinestring(TaggedPolygon.TaggedLineString ls) {
+        List<Object> result = new ArrayList<>();
+        List<TaggedPolygon.TaggedCoordinateSequence> coordinates = ls.getCoordinates();
+        int last = coordinates.size();
+        for (int i = 0; i < last; i++) {
+            TaggedPolygon.TaggedCoordinateSequence cs = coordinates.get(i);
+            Object value = buildTaggedCoordinateSequence(cs);
+            // client oddity: needs spaces before and after the map-span elements to work
```
Comment (aaime) [inline src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLGenerator.java:358]: Now that I think about it, I had noticed that space before/after span tag thing. But it's not working as intended in all cases, it seems.

Comment (aaime) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLGenerator.java:358]: Well duh, this change fixes 3 out of 4 of the cases above, only the countries one is left:

```
            if (value instanceof String) {
                if (i > 0) {
                    value = " " + value;
                } 
                if (i < last - 1) {
                    value = value + " ";
                }
            }
```

Comment (aaime) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLGenerator.java:358]: Fix pushed into this branch.

Comment (prushforth) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLGenerator.java:358]: > it's not working as intended in all cases, it seems.

No, there is definitely a bug in the viewer renderer. I am going to fix it, somehow.

**Comment by prushforth** [comment]
Locally fixed the rendering bug, with help from @aliyanh: 
image

Will work on getting our release out.

**Comment by prushforth** [comment]
I noticed when playing with the states layer that some polygons were serialized with no geometry, which was causing problems with the tile renderer.   I will try to fix that issue, but it might be good to eliminate map-features with no child map-geometry elements at source.

**Comment by aaime** [comment]
@prushforth yes, we definitely don't want to have emtpy geometries... I've already made some effort to clean out the empty ones, but I guess some are still going out of the gates. I'll see if I can spot them, and fix.

**Comment by prushforth** [comment]
To be clear, I tried the tiling features in the OSMTILE projection, and there were a few \<map-feature>\<map-properties>...\</map-properties> \</map-feature> i.e. no geometry at all, not just an empty geometry (which is a different test case now that I think of it).

**Comment by aaime** [comment]
I see. In those cases the feature should be skipped entirely. Going to check what's happening.

**Comment by prushforth** [comment]
> I've added a change to skip the features whose geometry is empty or missing entirely.
Are we good in terms of what the server is generating? If so, I'll "tighten the bolts" by adding more tests.

Yes, I belive so.  Those issues do look like client rendering problems.  I will look into it a bit, but I'm putting my efforts into getting the current viewer released and then to create a PR for GeoServer to update the viewer before the 2.25.0 release, if that is possible.

**Comment by prushforth** [comment]
> It's starting to feel a bit tight, isn't it?

I can imagine but I think you guys are doing great.  No worries from my POV.

**Comment by prushforth** [comment]
I noticed in the tasmania_cities layer when "Use features" is checked, I get a \<point> feature inside a map-geometrycollection. It should be map-point: 
image

**Comment by prushforth** [comment]
> Ok, with the last fix the tasmania layer should not be generating a geometry collection anymore.

I feel like I cheated because I didn't actually fix the viewer yet and it's working.  Thanks!

I think that when we send a \<map-link rel="tile" type="text/mapml">, we should a) also send a \<map-link rel="query" tref="...">, per the usual method.  We (Maps4HTML community group) don't yet have a good plan for how to deal with feature attributes in the client in a tiled-mapml situation, so they are just baggage and b) map-properties could be omitted from the response, opting for the query link as a lighter-weight option.

**Comment by aaime** [comment]
Closing, will create a new PR that's up to date with the current GeoServer code base (after the merge of filtering for vector mapml tiles)

**Pull request closed without merging**